File: sdks/typescript/src/documents.ts
/**
 * OpenRAG SDK documents client.
 */

import type { OpenRAGClient } from "./client";
import type { DeleteDocumentResponse, IngestResponse } from "./types";

export interface IngestOptions {
  /** Path to file (Node.js only). */
  filePath?: string;
  /** File object (browser or Node.js). */
  file?: File | Blob;
  /** Filename when providing file/blob. */
  filename?: string;
}

export class DocumentsClient {
  constructor(private client: OpenRAGClient) {}

  /**
   * Ingest a document into the knowledge base.
   *
   * @param options - Ingest options (filePath or file+filename).
   * @returns IngestResponse with document_id and chunk count.
   */
  async ingest(options: IngestOptions): Promise<IngestResponse> {
    const formData = new FormData();

    if (options.filePath) {
      // Node.js: read file from path
      if (typeof globalThis.process !== "undefined") {
        const fs = await import("fs");
        const path = await import("path");
        const fileBuffer = fs.readFileSync(options.filePath);
        const filename = path.basename(options.filePath);
        const blob = new Blob([fileBuffer]);
        formData.append("file", blob, filename);
      } else {
        throw new Error("filePath is only supported in Node.js");
      }
    } else if (options.file) {
      if (!options.filename) {
        throw new Error("filename is required when providing file");
      }
      formData.append("file", options.file, options.filename);
    } else {
      throw new Error("Either filePath or file must be provided");
    }

    const response = await this.client._request(
      "POST",
      "/api/v1/documents/ingest",
      {
        body: formData,
        isMultipart: true,
      }
    );

    const data = await response.json();
    return {
      success: data.success ?? false,
      document_id: data.document_id ?? null,
      filename: data.filename ?? null,
      chunks: data.chunks ?? 0,
    };
  }

  /**
   * Delete a document from the knowledge base.
   *
   * @param filename - Name of the file to delete.
   * @returns DeleteDocumentResponse with deleted chunk count.
   */
  async delete(filename: string): Promise<DeleteDocumentResponse> {
    const response = await this.client._request("DELETE", "/api/v1/documents", {
      body: JSON.stringify({ filename }),
    });

    const data = await response.json();
    return {
      success: data.success ?? false,
      deleted_chunks: data.deleted_chunks ?? 0,
    };
  }
}

File: sdks/python/openrag_sdk/documents.py
"""OpenRAG SDK documents client."""

from pathlib import Path
from typing import TYPE_CHECKING, BinaryIO

import httpx

from .models import DeleteDocumentResponse, IngestResponse

if TYPE_CHECKING:
    from .client import OpenRAGClient


class DocumentsClient:
    """Client for document operations."""

    def __init__(self, client: "OpenRAGClient"):
        self._client = client

    async def ingest(
        self,
        file_path: str | Path | None = None,
        *,
        file: BinaryIO | None = None,
        filename: str | None = None,
    ) -> IngestResponse:
        """
        Ingest a document into the knowledge base.

        Args:
            file_path: Path to the file to ingest.
            file: File-like object to ingest (alternative to file_path).
            filename: Filename to use when providing file object.

        Returns:
            IngestResponse with document_id and chunk count.

        Raises:
            ValueError: If neither file_path nor file is provided.
        """
        if file_path is not None:
            path = Path(file_path)
            with open(path, "rb") as f:
                files = {"file": (path.name, f)}
                response = await self._client._request(
                    "POST",
                    "/api/v1/documents/ingest",
                    files=files,
                )
        elif file is not None:
            if filename is None:
                raise ValueError("filename is required when providing file object")
            files = {"file": (filename, file)}
            response = await self._client._request(
                "POST",
                "/api/v1/documents/ingest",
                files=files,
            )
        else:
            raise ValueError("Either file_path or file must be provided")

        data = response.json()
        return IngestResponse(**data)

    async def delete(self, filename: str) -> DeleteDocumentResponse:
        """
        Delete a document from the knowledge base.

        Args:
            filename: Name of the file to delete.

        Returns:
            DeleteDocumentResponse with deleted chunk count.
        """
        response = await self._client._request(
            "DELETE",
            "/api/v1/documents",
            json={"filename": filename},
        )

        data = response.json()
        return DeleteDocumentResponse(**data)
